### DISCUSSION TITLE
Problem listing collections

### DISCUSSION THREAD

**Discussion by bskemp01** [ptarmiganlabs/butler-sheet-icons#468] [Q&A]
Hey all,

im trying to test the program and trying to list my collections in our tenant, however i am getting the following error:

![Capture](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/2e639951-1aee-4902-b73c-d90dfb963136)

any help to move forward from this would be appreciated :)

**Comment by mountaindude** [comment]
Well, that's a rather non-descriptive error message.... could certainly be improved.

Let me test things here to make sure Qlik hasn't recently changed something in QS Cloud.
I do test Butler Sheet Icons regularly to ensure it works against current QS Cloud version, it's been a couple of months since last check now though. 
Date for last check is found [here](https://github.com/ptarmiganlabs/butler-sheet-icons?tab=readme-ov-file#qlik-sense-cloud), btw.

**Comment by mountaindude** [comment]
Listing collections works for me:

<img width="1137" alt="image" src="https://github.com/ptarmiganlabs/butler-sheet-icons/assets/1029262/0b210653-6f5d-4a35-8cf1-b4579b2702ac">

My tenant is in EU rather than US. No idea whether QS Cloud differ behind the scenes between EU and US...

In theory two different tenants could be served different code by Qlik.
For example, when new versions of QS Cloud are rolled out that is probably done over several days (or longer?). 
No way to test that/verify that hypothesis though.

Maybe some of your parameters to BSI are incorrect?

In the example above that worked for me the variables had the following values:

```
BSI_CLOUD_TENANT_URL=hni......eu.qlikcloud.com
BSI_CLOUD_API_KEY=eyJhbGciOi.....
```

I do get the same error as you when I provide an incorrect or invalid API KEY.
So that could be worth looking into for you.

Meanwhile I'll put on the todo list to improve logging a bit in next version of the tool.

**Comment by bskemp01** [comment]
Good day,

Thank you for replying so quickly ☺️ I will check the API key but this
error was appearing for my colleague as well.

I will get back after I tested the key.

**Comment by bskemp01** [comment]
ok looks like its not the api key

![image](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/78c075f1-6b63-4785-b24c-afd1a187fe35)


i will keep trying some things

**Comment by mountaindude** [comment]
Hmm... Not sure what's going on there.

Could it be related to permissions for the user in question?
These things _have_ changed over past months in QS Cloud, maybe it's related.

Here are settings for my user:

<img width="1785" alt="image" src="https://github.com/ptarmiganlabs/butler-sheet-icons/assets/1029262/2ebde787-109f-4acd-8a4c-425b36b5b31e">


<img width="1775" alt="image" src="https://github.com/ptarmiganlabs/butler-sheet-icons/assets/1029262/ab12d467-5f42-46ee-8975-6f45e8c3a57d">


<img width="1778" alt="image" src="https://github.com/ptarmiganlabs/butler-sheet-icons/assets/1029262/591f1e0c-4b43-42fc-82eb-39db342024cd">

**Reply by bskemp01** [reply]
I am not the tenant owner but i am one of two developers here. i have all permissions you have except the steward.

**Comment by bskemp01** [comment]
i tried a different command the `create-sheet-thumbnails` and i see a little more info if this helps.

![image](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/94e6cd34-213f-4a93-af06-4f57f9a95d3b)

**Comment by mountaindude** [comment]
Same thing there, those `[object Object]` log messages aren't really helpful.

I am in the process of releasing a new BSI version with better/more helpful logging, should be out within the hour.
That will hopefully tell us more what's causing the problem.

You can also try adding `--loglevel verbose` or even `--loglevel debug` when running the tool. 
That will increase the logging and it may also give clues to what the root cause of the problem is.

**Comment by mountaindude** [comment]
@bskemp01, a new BSI version (3.2.2) has just been released, available [here](https://github.com/ptarmiganlabs/butler-sheet-icons/releases).

It's literally just minutes old so I have not tested it myself on Windows, but all the automated tests (incl the ones done on Windows) passed, so it should be fine.

That version includes better logging. 
Give it a spin and report back what you see, would be nice sorting out what's causing the issue you're seeing.

**Comment by bskemp01** [comment]
will do thank you!

**Comment by bskemp01** [comment]
ok well it does have a bit more information now. here is the screenshot along with the script i am using:

![image](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/4f3b6790-960a-4eed-97a9-a32d2246812c)

**Reply by mountaindude** [reply]
Ok, this gives us something to work with.

Looks like a call to QS Cloud fails and the result returned is not what's expected by BSI.
I'll see if I can reproduce the error based on this.

**Comment by bskemp01** [comment]
Great thanks! And take your time. I will not be back at work until Monday 😊

**Comment by mountaindude** [comment]
Think I may have found it.

You don't by any chance have "https://" at the beginning of the `--tenanturl` parameter?
When I have that I get the same error you get.

If that's the root cause it should prob be classified as a bug...
Formally a URL consists of a protocol (`https`) and a hostname (`tenantid.eu.qlikcloud.com`).
But Butler Sheet Icons expects only the hostname, i.e. that parameter is rather badly named.

Changing the name of the parameter would be a breaking change, I think a better option is to 
update BSI itself to gracefully accept either a full URL or a hostname, and then fix things so a proper call to QS Cloud can be made.

See bug #476.

**Comment by bskemp01** [comment]
Since this GitHub is my personal account I'm getting emails so I respond 😊
but no actually I did not have that. I have:

Tenant.us.qlikcloud.com

But that is an interesting bug.

**Comment by mountaindude** [comment]
Hmm. I was so sure.
Oh well, putting the thinking 🧢 back on then.

**Comment by bskemp01** [comment]
small update if it helps. i tried to use the node.js version and heres the error from that. not sure if its more info or not

![Capture](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/608a0147-4222-4f88-aac3-a5753864a590)

**Comment by mountaindude** [comment]
@bskemp01 

Version [3.3.0](https://github.com/ptarmiganlabs/butler-sheet-icons/releases/tag/butler-sheet-icons-v3.3.0) was released earlier today.
It adds both better/more relevant error logging, as well as a connection test to QS Cloud on startup.

All in all the new version should be better at telling what's causing the problems you ran into.

For example, if I list which QS Cloud collections are available to the user associated with the API key, it looks like this (in this case on macOS/Linux):

```bash
➜  butler-sheet-icons ./butler-sheet-icons qscloud list-collections --tenanturl $BSI_CLOUD_TENANT_URL --apikey $BSI_CLOUD_API_KEY --outputformat table
2024-02-16T15:27:01.951Z info: App version: 3.2.3
2024-02-16T15:27:01.952Z info: Testing connection to Qlik Sense Cloud...
2024-02-16T15:27:02.841Z info: Connection to tenant hnik......eu.qlikcloud.com successful.
2024-02-16T15:27:02.841Z info:     Tenant ID : cr0KDX....
2024-02-16T15:27:02.841Z info:     User name : Göran Sander
2024-02-16T15:27:02.842Z info:     User email: goran@....
2024-02-16T15:27:02.842Z info:     User ID   : 6337.....
2024-02-16T15:27:03.117Z info: Collections:
╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║                                                                   Butler Sheet Icons                                                                   ║
║                                            Available collections in tenant hni.............eu.qlikcloud.com                                            ║
╟───────────────────────┬────────────────────────────────┬──────────────────────────┬─────────┬────────────┬──────────────────────┬──────────────────────╢
║ Name                  │ Description                    │ ID                       │ Type    │ Item count │ Created at           │ Updated at           ║
╟───────────────────────┼────────────────────────────────┼──────────────────────────┼─────────┼────────────┼──────────────────────┼──────────────────────╢
║ Sheet thumbnails demo │                                │ 65cb8b.................. │ private │ 1          │ 2024-02-13T15:33:08Z │ 2024-02-13T15:36:28Z ║
╚═══════════════════════╧════════════════════════════════╧══════════════════════════╧═════════╧════════════╧══════════════════════╧══════════════════════╝

➜  butler-sheet-icons
```

Now, if I run the same command with an invalid or expired API token, I will with the latest version of Buter Sheet Icons get more informative errors compared to earlier versions:


```bash
➜  butler-sheet-icons ./butler-sheet-icons qscloud list-collections --tenanturl $BSI_CLOUD_TENANT_URL --apikey an-invalid-key --outputformat table
2024-02-16T15:30:54.510Z info: App version: 3.2.3
2024-02-16T15:30:54.511Z info: Testing connection to Qlik Sense Cloud...
2024-02-16T15:30:55.187Z error: LIST COLLECTIONS 1 (message): Request failed with status code 401
2024-02-16T15:30:55.187Z error: LIST COLLECTIONS 1 (error code): 401="Unauthorized"
➜  butler-sheet-icons
```
 
Which makes sense, ´an-invalid-key` is obviously not a proper/valid API token.

**Comment by bskemp01** [comment]
ok. with the updates to the programs here is my new screenshot. i hope this information helps you

![image](https://github.com/ptarmiganlabs/butler-sheet-icons/assets/93053776/e8e35106-5c64-4574-ae57-7909a12eb08d)
